packages/core-components/src/layout/DelegatedSignInPage/DelegatedSignInIdentity.ts

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  BackstageUserIdentity,
+  discoveryApiRef,
+  fetchApiRef,
+  IdentityApi,
+  ProfileInfo,
+} from '@backstage/core-plugin-api';
+import { ResponseError } from '@backstage/errors';
+import { DelegatedSession, delegatedSessionSchema } from './types';
+
+export const DEFAULTS = {
+  // The amount of time between token refreshes, if we fail to get an actual
+  // value out of the exp claim
+  defaultTokenExpiryMillis: 60_000,
+  // The amount of time before the actual expiry of the Backstage token, that we
+  // shall start trying to get a new one
+  tokenExpiryMarginMillis: 30_000,
+} as const;
+
+// When the token expires, with some margin
+export function tokenToExpiry(jwtToken: string | undefined): Date {
+  if (!jwtToken) {
+    return new Date(Date.now() + DEFAULTS.defaultTokenExpiryMillis);
+  }
+
+  const [_header, rawPayload, _signature] = jwtToken.split('.');
+  const payload = JSON.parse(atob(rawPayload));
+
+  return new Date(payload.exp * 1000 - DEFAULTS.tokenExpiryMarginMillis);
```

Comment:
Perhaps do a `payload.exp` check here and fall back to default if not set for safety?